luarules/gadgets/unit_custom_weapons_cluster.lua

```diff
@@ -59,281 +58,258 @@ local cos   = math.cos
+					local cosy = cos(ph_y)
+					dx = dx + bounce * sin(th_z) * cosy
+					dy = dy + bounce * sin(ph_y)
+					dz = dz + bounce * cos(th_z) * cosy
+				end
+			end
+		end
+	end
+
+	return dx, dy, dz
 end
 
-local function SpawnClusterProjectiles(data, attackerID, ex, ey, ez, deflection)
-    local projNum = data.number
-    local projVel = data.projVel
-
-    spawnCache.owner = attackerID or -1
-    spawnCache.ttl   = data.projTtl
-
-    -- Initial direction vectors are evenly spaced.
-    local directions = directions[projNum]
-
-    local vx, vy, vz, norm
-    for ii = 0, (projNum-1) do
-        -- Avoid shooting into terrain by adding deflection.
-        vx = directions[3*ii+1] + deflection[1]
-        vy = directions[3*ii+2] + deflection[2]
-        vz = directions[3*ii+3] + deflection[3]
-
-        -- Since the initial directions are not random, we add jitter.
-        -- Note: Comment this out to test without any randomness.
-        vx = vx + (rand() * 6 - 3) / projNum
-        vy = vy + (rand() * 6 - 3) / projNum
-        vx = vx + (rand() * 6 - 3) / projNum
-
-        -- Set the projectile's velocity vector.
-        norm = sqrt(vx*vx + vy*vy + vz*vz)
-        vx = vx * projVel / norm
-        vy = vy * projVel / norm
-        vz = vz * projVel / norm
-        spawnCache.speed = { vx, vy, vz }
-
-        -- Pre-scatter the projectile and set its initial position.
-        spawnCache.pos = {
-            ex + vx * GAME_SPEED / 2,
-            ey + vy * GAME_SPEED / 10 * max(1, 5 * abs(vy) / projVel),
-            ez + vz * GAME_SPEED / 2
-        }
-
-        spSpawnProjectile(data.projDef, spawnCache)
-    end
+local function SpawnClusterProjectiles(data, projectileID, attackerID, ex, ey, ez)
+	local clusterDefID = data.weaponID
+	local projectileCount = data.number
+	local projectileSpeed = data.weaponSpeed
+
+	spawnCache.owner = attackerID or -1
+	spawnCache.ttl = data.weaponTtl
+	local speed = spawnCache.speed
+	local pos = spawnCache.pos
+
+	local directions = directions[projectileCount]
+	local deflectX, deflectY, deflectZ = GetSurfaceDeflection(data, projectileID, ex, ey, ez)
+	local randomness = 1 / sqrt(projectileCount - 2)
+
+	for ii = 0, projectileCount - 1 do
+		local vx = directions[3*ii+1] + deflectX
+		local vy = directions[3*ii+2] + deflectY
+		local vz = directions[3*ii+3] + deflectZ
+
+		vx = vx + (rand() - 0.5) * randomness * 2
+		vy = vy + (rand() - 0.5) * randomness * 2
+		vz = vz + (rand() - 0.5) * randomness * 2
+
+		-- Higher projectile counts will have less variation in projectile speed.
+		local normalization = (1 + rand() * randomness) / (1 + randomness)
+		normalization = normalization * projectileSpeed / sqrt(vx*vx + vy*vy + vz*vz)
+		vx = vx * normalization
+		vy = vy * normalization
+		vz = vz * normalization
+
+		speed[1] = vx
+		speed[2] = vy
+		speed[3] = vz
+
+		pos[1] = ex + vx * gameSpeed / 2
+		pos[2] = ey + vy * gameSpeed / 2
+		pos[3] = ez + vz * gameSpeed / 2
```

Comment:
Why /2?

Reply:
Anywhere you see a magic number, my answer is 🤷❓